### ISSUE TITLE
Load Test failed incase more than 7 requests at the same time

### ISSUE THREAD

**Issue by dungle01** [entrpn/serving-diffusion#5]
Hi,

I deployed to Vertex AI with 1 compute node, "n1-highmem-4" and 'NVIDIA_TESLA_V100'. 

Then I did a load test with Jmeter with Number of Threads: 8, Loop count: 1 ( 8 requests at the same time ), the result is 100% failed. It's 100% success with only 6 requests at the same time

I saw that it failed at this PLMSSampler step in txt2img.py :
                            samples_ddim, _ = sampler.sample(S=ddim_steps,
                                                            conditioning=c,
                                                            batch_size=n_samples,
                                                            shape=shape,
                                                            verbose=False,
                                                            unconditional_guidance_scale=scale,
                                                            unconditional_conditioning=uc,
                                                            eta=ddim_eta,
                                                            x_t=start_code
                                                            )

Regards

**Comment by entrpn** [comment]
@dungle01 Thanks for running this test. Its good to know what a single replica can handle concurrently.

If you like to have more concurrency, you'll need start with a [min_replica_count](https://github.com/entrpn/serving-diffusion/blob/master/gcp_deploy.py#L11) higher than 1 and you can set the [max_replica_count](https://github.com/entrpn/serving-diffusion/blob/master/gcp_deploy.py#L12) to a higher value as well.

I'm working on making some changes to this script so that its configurable via parameters.

**Comment by dungle01** [comment]
@entrpn the replica-count is used for autoscaling with CPU metrics. But this case is not about high CPU, it seem PLMSSampler won't work well with many concurrently run in a same VM.

Vertex AI won't distribute the traffic into new VM if the first VM still have low CPU metric.

Regards

**Comment by mln-wave** [comment]
@dungle01 
But as per what is mentioned [here](https://cloud.google.com/vertex-ai/docs/predictions/configure-compute#scaling), we have an option to specify to autoscale on basis of gpu.
And I found the code/argument to mention this metric_spec in python sdk [here](https://github.com/googleapis/python-aiplatform/blob/cc7c968512db84385c053a8353a8dad8a57aef22/google/cloud/aiplatform/models.py#L804).

**Comment by entrpn** [comment]
I wrote more information about this issue [here](https://github.com/entrpn/serving-diffusion/issues/7#issuecomment-1251173939).

I'll be running some experiments this week and get back to you with the results.

**Comment by entrpn** [comment]
@dungle01 as @mln-wave mentioned, autoscaling is also based on GPU usage. However, autoscaling is slow, so is better to start with more replicas. You can run the deploy command with something like.

```python
python gcp_deploy.py --image-uri gcr.io/<project_id>/stable-diffusion:latest --max-replica-count 5 --min-replica-count 5
```

The script has also been updated to replicate with a lower accelerator duty cycle.

**Comment by dungle01** [comment]
thanks @entrpn , 

let me double-check on our side

**Comment by dungle01** [comment]
@entrpn after I deployed latest code, I got the image data but it's not an image (see attachments). 
Did you got it on your side ?
<img width="1125" alt="Screen Shot 2022-09-21 at 11 12 36" src="https://user-images.githubusercontent.com/91128667/191413015-484755fa-a60c-4f27-bcea-2b4530b45cde.png">
<img width="1301" alt="Screen Shot 2022-09-21 at 11 12 52" src="https://user-images.githubusercontent.com/91128667/191413021-25c339f2-1821-4305-bfbe-30c55e4d96ea.png">

**Comment by entrpn** [comment]
@dungle01 I am not running into this issue. I don't think you are using the latest codebase since the response object has changed to something like:

```
{"predictions":[{"prompt":"An apocalyptic alien city ravaged by time, highly detailed, cinematic, intricate, trending in artstation","images":["base64encodedimage","base64encodedimage"]}]}
```

The request has also changed to include the parameters within the instances list. Ex:

```
{"instances": [{"prompt": "An apocalyptic alien city ravaged by time, highly detailed, cinematic, intricate, trending in artstation", "parameters": {"ddim_steps": 30, "scale": 7.5, "n_samples": 2, "n_itter": 2, "type": "txt2img"}}]}
```

Inside the stable-diffusion folder there is a python script `validate_response.py` that you can run that will take the response.json and decode the base64 images.

**Comment by dungle01** [comment]
@entrpn 

I can managed to update latest code now, there's some wrong cache from docker build on my side.

For Load Test checking:
- With test min=4, max=4 : I saw it can handle well with 10 reqs at a same time, up to 20 reqs with  low rate of error occassionally
- With test min=1, max=4: appear high rate of error (sometime 90%) when reach 10 reqs. It seems that it work not well in scaling gap.

A work-around that I implemented a job queue to control the max no of concurrent reqs (~10reqs) and use min=4, max=4 mode. The job queue is implemented by Cloud Function with Firebase FireStore .
